librz/analysis/var_global.c

```diff
@@ -0,0 +1,480 @@
+// SPDX-FileCopyrightText: 2021 Basstorm <basstorm@nyist.edu.cn>
+// SPDX-License-Identifier: LGPL-3.0-only
+
+#include <rz_analysis.h>
+#include <rz_list.h>
+
+/**
+ * \brief Create a new instance of global variable
+ * 
+ * \param name variable name
+ * \param addr variable address
+ * \param comment variable comment
+ * \return RzAnalysisVarGlobal *
+ */
+RZ_API RZ_OWN RzAnalysisVarGlobal *rz_analysis_var_global_new(RZ_NONNULL const char *name, ut64 addr, RZ_NULLABLE const char *comment) {
+	rz_return_val_if_fail(name && addr, NULL);
+	RzAnalysisVarGlobal *glob = RZ_NEW0(RzAnalysisVarGlobal);
+	if (!glob) {
+		return NULL;
+	}
+	glob->name = strdup(name);
+	glob->addr = addr;
+	if (comment) {
+		glob->comment = strdup(comment);
+	}
+	return glob;
+}
+
+typedef struct {
+	RBNode rb;
+	ut64 addr;
+	ut64 size;
+} GlobalVarNode;
+
+int global_var_node_cmp(const void *incoming, const RBNode *in_tree, void *user) {
+	ut64 ia = *(ut64 *)incoming;
+	ut64 ta = container_of(in_tree, const GlobalVarNode, rb)->addr;
+	if (ia < ta) {
+		return -1;
+	} else if (ia > ta) {
+		return 1;
+	}
+	return 0;
+}
+
+/**
+ * \brief Add the global variable into hashtable
+ * 
+ * \param analysis RzAnalysis
+ * \param global_var Global variable instance
+ * \return true if succeed
+ */
+RZ_API RZ_OWN bool rz_analysis_var_global_add(RzAnalysis *analysis, RzAnalysisVarGlobal *global_var) {
+	rz_return_val_if_fail(analysis && global_var, false);
+	if (rz_analysis_var_global_get_byaddr(analysis, global_var->addr)) {
+		RZ_LOG_ERROR("Global variable %s at 0x%" PFMT64x " already exists!\n", global_var->name, global_var->addr);
+		return false;
+	} else if (ht_pp_find(analysis->ht_global_var, global_var->name, NULL)) {
+		RZ_LOG_ERROR("Global variable %s already exists!\n", global_var->name);
+		return false;
+	}
+	if (!ht_pp_insert(analysis->ht_global_var, global_var->name, global_var)) {
+		return false;
+	}
+	GlobalVarNode *node = RZ_NEW0(GlobalVarNode);
+	node->addr = global_var->addr;
+	node->size = 0;
+	if (global_var->type) {
+		node->size = rz_type_db_get_bitsize(analysis->typedb, global_var->type) / 8;
+	}
+	if (!rz_rbtree_aug_insert(&analysis->global_var_tree, &node->addr, &node->rb, global_var_node_cmp, NULL, NULL)) {
```

Comment:
Do we need an augmented RBTree? cc @thestr4ng3r

Reply:
No, augmentation is only useful when overlaps are allowed, but for global variables this wouldn't make much sense. Instead of `rz_rbtree_aug_insert`, `rz_rbtree_insert` can be used here (but it's just an alias).